session/schema_amender.go

```diff
@@ -0,0 +1,563 @@
+}
+
+func (a *amendOperationAddIndexInfo) genIndexKeyValue(ctx context.Context, sctx sessionctx.Context, kvMap map[string][]byte,
+	key []byte, kvHandle kv.Handle, keyOnly bool) ([]byte, []byte, error) {
+	colMap := make(map[int64]*types.FieldType)
+	for _, oldCol := range a.tblInfoAtStart.Meta().Cols() {
+		colMap[oldCol.ID] = &oldCol.FieldType
+	}
+	chk := a.chk
+	chk.Reset()
+	// The Op_Put may not exist in old value kv map.
+	val, ok := kvMap[string(key)]
+	if !ok && keyOnly {
+		return nil, nil, nil
+	}
+	err := executor.DecodeRowValToChunk(sctx, a.schemaAndDecoder.schema, a.tblInfoAtStart.Meta(), kvHandle, val, chk, a.schemaAndDecoder.decoder)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend decode value to chunk failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	idxVals := make([]types.Datum, 0, len(a.indexInfoAtCommit.Meta().Columns))
+	for _, oldCol := range a.relatedOldIdxCols {
+		idxVals = append(idxVals, chk.GetRow(0).GetDatum(oldCol.Offset, &oldCol.FieldType))
+	}
+
+	// Generate index key buf.
+	newIdxKey, distinct, err := tablecodec.GenIndexKey(sctx.GetSessionVars().StmtCtx,
+		a.tblInfoAtCommit.Meta(), a.indexInfoAtCommit.Meta(), a.tblInfoAtCommit.Meta().ID, idxVals, kvHandle, nil)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend generate index key failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	if keyOnly {
+		return newIdxKey, []byte{}, nil
+	}
+
+	// Generate index value buf.
+	var containsNonBinaryString bool
+	for _, idxCol := range a.indexInfoAtCommit.Meta().Columns {
+		col := a.tblInfoAtCommit.Meta().Columns[idxCol.Offset]
+		if col.EvalType() == types.ETString && !mysql.HasBinaryFlag(col.Flag) {
+			containsNonBinaryString = true
+			break
+		}
+	}
+	newIdxVal, err := tablecodec.GenIndexValue(sctx.GetSessionVars().StmtCtx, a.tblInfoAtCommit.Meta(),
+		a.indexInfoAtCommit.Meta(), containsNonBinaryString, distinct, false, idxVals, kvHandle)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend generate index values failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	return newIdxKey, newIdxVal, nil
+}
+
+func (a *amendOperationAddNewIndex) processKey(ctx context.Context, sctx sessionctx.Context, key []byte,
+	kvMap map[string][]byte) (resAdd tikv.CommitterMutations, err error) {
+	kvHandle, err := tablecodec.DecodeRowKey(key)
+	if err != nil {
+		logutil.Logger(ctx).Error("decode key error", zap.String("key", hex.EncodeToString(key)), zap.Error(err))
+		return resAdd, errors.Trace(err)
+	}
+
+	newIdxKey, newIdxValue, err := a.info.genIndexKeyValue(ctx, sctx, kvMap, key, kvHandle, false)
+	if err != nil {
+		return resAdd, errors.Trace(err)
+	}
+	// Check if the generated index keys are unique for unique index.
+	newIndexOp := pb.Op_Put
+	if a.info.indexInfoAtCommit.Meta().Unique {
+		if _, ok := a.processedNewIndexKeys[string(newIdxKey)]; ok {
+			return resAdd, errors.Trace(errors.Errorf("amend process key same key=%v found for unique index=%v",
```

Comment:
amend unique index logic is complex, can we remove the related logic in this already large PR?.